src/main.rs

```diff
@@ -0,0 +1,116 @@
+extern crate simple_proxy;
+
+mod aws_signature_builder;
+
+use simple_proxy::{SimpleProxy, Environment};
+
+use structopt::StructOpt;
+
+#[derive(StructOpt, Debug)]
+struct Cli {
+    port: u16,
+    region: String,
+}
+
+use rusoto_credential::{ProvideAwsCredentials, DefaultCredentialsProvider};
+use futures::future::Future;
+
+use hyper::header::HeaderValue;
+use hyper::{Body, Request};
+
+use hyper::header::AUTHORIZATION;
+
+const XAMZCONTENTSHA256: &str = "x-amz-content-sha256";
+const XAMZSECURITYTOKEN: &str = "x-amz-security-token";
+const XAMZDATE: &str = "x-amz-date";
+
+use simple_proxy::proxy::error::MiddlewareError;
+use simple_proxy::proxy::middleware::MiddlewareResult::Next;
+use simple_proxy::proxy::middleware::{Middleware, MiddlewareResult};
+use simple_proxy::proxy::service::{ServiceContext, State};
+
+pub struct AwsSignatureHeaders {
+    region: String
+}
+
+impl AwsSignatureHeaders {
+    pub fn new(region: String) -> Self {
+        AwsSignatureHeaders{
+            region: region
+        }
+    }
+}
+
+impl Middleware for AwsSignatureHeaders {
+    fn name() -> String {
+        String::from("AwsSignatureHeaders")
+    }
+
+    fn before_request(
+        &mut self,
+        req: &mut Request<Body>,
+        _context: &ServiceContext,
+        _state: &State,
+    ) -> Result<MiddlewareResult, MiddlewareError> {
+        let (amzdate, datestamp) = aws_signature_builder::get_utc_aws_signature_datestrings();
+        let mut headers = ::std::collections::HashMap::new();
+        for (key, value) in req.headers().iter() {
+            headers.insert(String::from(key.as_str()), String::from(value.to_str().unwrap()));
+        }
```

Comment:
This is quite bad, and really comes from the fact that the function takes the headers as a hashmap from string to string (mainly because it was a python dict before).

I like that the library has no dependencies, but maybe no one really cares about that.

Hmm, also now that I'm reading [the HeaderMap docs](https://hyper.rs/hyper/master/hyper/header/index.html#headermap), I can see:

> It is a multimap structure, where each header name may have multiple associated header values. Given this, some of the APIs diverge from HashMap.

That might be a good reason to just take a HeaderMap.